tests/draft6/ref.json

```diff
@@ -328,5 +328,110 @@
                 "valid": false
             }
         ]
+    },
+    {
+        "description": "Location-independent identifier",
+        "schema": {
+            "$ref": "#foo",
+            "definitions": {
+                "A": {
+                    "$id": "#foo",
+                    "type": "integer"
+                }
+            }
+        },
+        "tests": [
+            {
+                "data": 1,
+                "description": "match",
+                "valid": true
+            },
+            {
+                "data": "a",
+                "description": "mismatch",
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "Canonical location-independent identifier",
+        "schema": {
+            "$ref": "http://localhost:1234/bar#foo",
+            "definitions": {
+                "A": {
+                    "$id": "http://localhost:1234/bar#foo",
+                    "type": "integer"
+                }
+            }
+        },
+        "tests": [
+            {
+                "data": 1,
+                "description": "match",
+                "valid": true
+            },
+            {
+                "data": "a",
+                "description": "mismatch",
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "Location-independent identifier with base URI change in subschema",
+        "schema": {
+            "$id": "http://localhost:1234/root",
+            "$ref": "http://localhost:1234/nested.json#foo",
+            "definitions": {
+                "A": {
+                    "$id": "nested.json",
+                    "definitions": {
+                        "B": {
+                            "$id": "#foo",
+                            "type": "integer"
+                        }
+                    }
+                }
+            }
+        },
+        "tests": [
+            {
+                "data": 1,
+                "description": "match",
+                "valid": true
+            },
+            {
+                "data": "a",
+                "description": "mismatch",
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "Location-independent identifiers should take precedent over regular fragments",
+        "schema": {
+            "$ref": "#foo",
```

Comment:
I really want the `$ref`s to all be in `allOf`s, I've decided.  It's such a confusing and underspecified area, when you then start throwing in odd corner cases like an unrecognized "foo" keyword alongside a `$ref` I have absolutely no idea what _should_ happen, and even less clue what the majority of implementations will do.

Let's avoid gray areas except when explicitly testing around them.